dcrrates/server/types.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// NewRateServer is a constructor for a RateServer.
+func NewRateServer(index string, xcBot *exchanges.ExchangeBot) *RateServer {
+	return &RateServer{
+		btcIndex:   index,
+		clientLock: new(sync.RWMutex),
+		clients:    make(map[StreamID]RateClient),
+		xcBot:      xcBot,
+	}
+}
+
+// GRPCStream wraps the grpc.ClientStream.
+type GRPCStream interface {
+	Send(*dcrrates.ExchangeRateUpdate) error
+	Context() context.Context
+}
+
+// sendStateList is a helper for parsing the ExchangeBotState when a new client
+// subscription is received.
+func sendStateList(client RateClient, updates map[string]*exchanges.ExchangeState) (err error) {
+	for token, update := range updates {
+		err = client.SendExchangeUpdate(makeExchangeUpdate(token, update))
+		if err != nil {
+			log.Errorf("SendExchangeUpdate error for %s: %v", token, err)
+			return
+		}
+	}
+	return
+}
+
+// SubscribeExchanges is a gRPC method defined in dcrrates/dcrrates.proto. It
+// satisfies the DCRRatesServer interface. The gRPC server will call this method
+// when a client subscription is received.
+func (server *RateServer) SubscribeExchanges(hello *dcrrates.ExchangeSubscription, stream dcrrates.DCRRates_SubscribeExchangesServer) (err error) {
+	return server.ReallySubscribeExchanges(hello, stream)
+}
+
+// ReallySubscribeExchanges stores the client and their exchange subscriptions.
+// It waits for the client contexts Done channel to close and deletes the client
+// from the RateServer.clients map.
+func (server *RateServer) ReallySubscribeExchanges(hello *dcrrates.ExchangeSubscription, stream GRPCStream) (err error) {
+	// For now, require the ExhcnageBot clients to have the same base currency.
+	// ToDo: Allow any index.
+	if hello.BtcIndex != server.btcIndex {
+		return fmt.Errorf("Exchange subscription has wrong BTC index. Given: %s, Required: %s", hello.BtcIndex, server.btcIndex)
+	}
+	// Save the client for use in the main loop.
+	client, sid := server.AddClient(stream, hello)
+	state := server.xcBot.State()
+	// Send Decred exchanges.
+	err = sendStateList(client, state.DcrBtc)
+	if err != nil {
+		return err
+	}
+	// Send Bitcoin-fiat indices.
+	for token := range state.FiatIndices {
+		client.SendExchangeUpdate(&dcrrates.ExchangeRateUpdate{
+			Token:   token,
+			Indices: server.xcBot.Indices(token),
+		})
+	}
+
+	// Get the address for an Infof. Seems like peerInfo is always found, but
+	// am checking anyway.
+	peerInfo, peerFound := grpcPeer.FromContext(client.Stream().Context())
+	if peerFound {
+		log.Infof("Client has connected from %s", peerInfo.Addr.String())
+	}
+
+	// Keep stream alive.
+	<-client.Stream().Context().Done()
+
+	server.DeleteClient(sid)
+	if peerFound {
+		log.Infof("Client at %s has disconnected", peerInfo.Addr.String())
+	}
+	return
+}
+
+// AddClient adds a client to the map and advance the streamCounter.
+func (server *RateServer) AddClient(stream GRPCStream, hello *dcrrates.ExchangeSubscription) (RateClient, StreamID) {
```

Comment:
Seems like AddClient and DeleteClient have no use outside of `dcrrates/server`.  Should/could they be unexported?